playbook/actions/ai.go

```diff
@@ -0,0 +1,277 @@
+package actions
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/context"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/duty/query"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/llm"
+	"github.com/samber/lo"
+)
+
+type promptContext struct {
+	config          *models.ConfigItem
+	relatedConfigs  []query.RelatedConfig
+	relatedChanges  []query.ConfigChangeRow
+	analysisResults []models.ConfigAnalysis
+}
+
+type AIAction struct{}
+
+type AIActionResult struct {
+	Logs string `json:"logs,omitempty"` // TODO: only naming this "logs" because the frontend has proper formatted display for this field
+}
+
+func (t *AIAction) Run(ctx context.Context, spec v1.AIAction) (*AIActionResult, error) {
+	if spec.Backend == "" {
+		spec.Backend = api.LLMBackendOpenAI
+	}
+
+	if apiKey, err := ctx.GetEnvValueFromCache(spec.APIKey, ctx.GetNamespace()); err != nil {
+		return nil, err
+	} else {
+		spec.APIKey.ValueStatic = apiKey
+	}
+
+	prompt, err := buildPrompt(ctx, spec.Prompt, spec.AIActionContext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to form prompt: %w", err)
+	}
+
+	llmConf := llm.Config{AIActionClient: spec.AIActionClient, UseAgent: spec.UseAgent}
+	response, err := llm.Prompt(ctx, llmConf, prompt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AIActionResult{Logs: response}, nil
+}
+
+func buildPrompt(ctx context.Context, prompt string, spec v1.AIActionContext) (string, error) {
+	pctx, err := getPromptContext(ctx, spec)
+	if err != nil {
+		return "", fmt.Errorf("failed to get prompt context: %w", err)
+	}
+
+	paragraphs := []string{
```

Comment:
This should be formatted into JSON:

```yaml
models.ConfigItem
config: json.RawMessage
changes: []models.ConfigChange
analysis: []models.ConfigInsight
graph: []struct{
{from, to, type, name}
},
```

Reply:
```
	[]llms.MessageContent{
			{
				Role: llms.ChatMessageTypeHuman,
				Parts: []llms.ContentPart{

					llms.BinaryPart("application/json", json),
					llms.TextPart("Please tell me the string on the box."),
				},
			},
		},
```

Reply:
why do we want to send this as json? 
what we currently have is more concise and shows the relationship explicitly.

Reply:
```go
llms.BinaryPart("application/json", json)
```

this is not allowed